File: prototype/node/src/app/core/customers/customer-merge-plan.ts
import type { ListingId } from '../ids/entity-ids.ts'

export type CartLine = { listingId: ListingId; quantity: number }

export type CustomerMergePlan = {
  /** The verified customer's cart after the fold, one line per listing. */
  cartLines: readonly CartLine[]
  /** Anonymous favorites the verified customer does not already have — repoint these rows. */
  favoritesToMove: readonly ListingId[]
  /** Anonymous favorites that duplicate one the verified customer already has — delete these rows. */
  favoritesToDrop: readonly ListingId[]
}

function foldCartLines(
  verifiedLines: readonly CartLine[],
  anonymousLines: readonly CartLine[],
  stockByListing: ReadonlyMap<ListingId, number>,
): CartLine[] {
  const order: ListingId[] = []
  const quantityByListing = new Map<ListingId, number>()

  for (const line of [...verifiedLines, ...anonymousLines]) {
    if (!quantityByListing.has(line.listingId)) {
      order.push(line.listingId)
    }
    quantityByListing.set(line.listingId, (quantityByListing.get(line.listingId) ?? 0) + line.quantity)
  }

  const lines: CartLine[] = []
  for (const listingId of order) {
    const summed = quantityByListing.get(listingId) ?? 0
    const stock = stockByListing.get(listingId)
    const quantity = stock === undefined ? summed : Math.min(summed, Math.max(stock, 0))
    if (quantity > 0) {
      lines.push({ listingId, quantity })
    }
  }
  return lines
}

/**
 * Splits the anonymous customer's favorites into the ones that can move
 * (nothing named that listing yet) and the ones that must be dropped instead
 * (the verified customer already favorited it, so moving the row would
 * duplicate it).
 */
function partitionFavorites(
  verifiedIds: readonly ListingId[],
  anonymousIds: readonly ListingId[],
): Pick<CustomerMergePlan, 'favoritesToMove' | 'favoritesToDrop'> {
  const alreadyFavorited = new Set(verifiedIds)
  const seen = new Set<ListingId>()
  const favoritesToMove: ListingId[] = []
  const favoritesToDrop: ListingId[] = []

  for (const listingId of anonymousIds) {
    if (seen.has(listingId)) continue
    seen.add(listingId)

    if (alreadyFavorited.has(listingId)) {
      favoritesToDrop.push(listingId)
    } else {
      favoritesToMove.push(listingId)
    }
  }

  return { favoritesToMove, favoritesToDrop }
}

export function planCustomerMerge(input: {
  verifiedCartLines: readonly CartLine[]
  anonymousCartLines: readonly CartLine[]
  verifiedFavoriteListingIds: readonly ListingId[]
  anonymousFavoriteListingIds: readonly ListingId[]
  /** Units in stock per listing. A listing absent from this map contributes no cap. */
  stockByListing: ReadonlyMap<ListingId, number>
}): CustomerMergePlan {
  return {
    cartLines: foldCartLines(input.verifiedCartLines, input.anonymousCartLines, input.stockByListing),
    ...partitionFavorites(input.verifiedFavoriteListingIds, input.anonymousFavoriteListingIds),
  }
}

File: prototype/rails/src/app/models/customer_merge_plan.rb
# The arithmetic a customer merge folds two carts and two favorites lists
# through: cart quantities summed per listing, clamped to what stock is left,
# a line that clamps to zero dropped; favorites de-duplicated by listing, the
# verified customer's own favorite kept over a duplicate. Pure math over plain
# data — no database access — so the fold's arithmetic is tested without a
# customer or a listing existing anywhere. `Customer#absorb` is what turns the
# answer into rows.
class CustomerMergePlan
  CartLine = Data.define(:listing_id, :quantity)

  # +verified+ and +anonymous+ are Hash(listing_id => quantity). +stock+ is
  # Hash(listing_id => units left); a listing absent from it contributes no
  # cap.
  def self.fold_cart_lines(verified, anonymous, stock)
    verified.merge(anonymous) { |_listing_id, from_verified, from_anonymous| from_verified + from_anonymous }
      .filter_map do |listing_id, quantity|
        cap = stock[listing_id]
        clamped = cap.nil? ? quantity : [ quantity, [ cap, 0 ].max ].min
        CartLine.new(listing_id: listing_id, quantity: clamped) if clamped.positive?
      end
  end

  # Anonymous favorites the verified customer does not already hold move;
  # ones that duplicate a favorite the verified customer already holds drop
  # instead of colliding with the unique index on (customer_id, listing_id).
  def self.partition_favorites(verified_listing_ids, anonymous_listing_ids)
    already_favorited = verified_listing_ids.to_set
    move, drop = anonymous_listing_ids.uniq.partition { |listing_id| already_favorited.exclude?(listing_id) }

    { move: move, drop: drop }
  end
end
